[WellSQL Migration] Migrate `WCGlobalAttributeModel` to Room

ParaskP7 opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Closes: [AINFRA-552](https://linear.app/a8c/issue/AINFRA-552)
Relates To: #14453

---

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCGlobalAttributeModel` from [WellSQL](https://github.com/wordpress-mobile/wellsql) to [Room](https://developer.android.com/training/data-storage/room) database.

FYI: This is a follow-up to #14453 and the removal of `WCAttributeTermModel`. This follow-up also removes the `termsId` column from the `WCGlobalAttributeModel` table before the actual migration starts (6e0700759fae7613cdd324be125c146bf5383cf2).

> I actually lost more time trying to connect the dots rather than migrating/removing this table... 😭

I now have a hypothesis about whether this `WCAttributeTermModel` table, the `termsId` column from the `WCGlobalAttributeModel` table, and all the unnecessary logic within the corresponding Store/Rest/SQL classes (which btw, now got removed: 2f1b0e7ddfd8ecef0c017c6599cc7a3596e769bf). My hypothesis is, as part of this [FluxC#1803](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1803) PR, especially after seeing [this](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1803/commits/b0f07ae9254b02916ae18d6d0462df524a394c13) commit, Thomaz made all attribute related CRUD operations available no matter if there were to be used (or not). Most probably the plan was to use them at some point, but, that never actually happened. Thus, all this logic ended-up being unused, and at best only used within FluxC's [example](https://github.com/wordpress-mobile/WordPress-FluxC-Android/tree/trunk/example) app.

---

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

1. Go to the `Products` screen.
2. Find a `Variable product` or change a `Physical product` to a `Variable product`.
3. Click on `Add variations` to navigate to the `Add attribute` screen.
4. Tap to select an existing attribute (like `Color` or `Size`) and see the list of option of that attribute.
5. Tap on a couple of options and then click `NEXT` twice, then press the `GENERATE VARIATION` button and select `Generate all variations`, following by `OK` when the dialog appears.
6. Click on `Add variations` again to navigate to the `Variations` screen where you'll see the variations you added.
7. Go back and click the `SAVE` button to have these changes apply to the product.

`EXTRA`

You might also want to add a new attribute on your store and see that everything is working as expected. To do so, follow the below instructions:

1. Go to your store's `wp-admin`.
2. Find the `Products` section, and within it, click on the `Attributes` sub-section.
3. Add a new attribute, following the instructions. For example, you could add a new attribute named `Material` with the `the-material-of-the-product` slug.
4. Click on `Configure terms` for that newly added attributed on that table, and then add a few terms for that attribute, following the instructions. For example, you could add a new term named `Cotton` with the `cotton-product-material` slug, etc. Add a few of those.
5. Now, open the mobile app and follow the above instructions once more. Making sure that this newly added attribute, along with all its terms are shown and usable.

---

### Images/gif

<img width="1160" height="624" alt="image" src="https://github.com/user-attachments/assets/a9f95df6-0776-4021-a683-9a95d68f631e" />

---

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Db: Remove unnecessary 'termsId' from 'WCGlobalAttributeModel' table

FYI: Similarly to this 364c18dcb07be26654f80d6c927ac0a03b4ccabb change,
which completely removed everything 'WCAttributeTermModel', this related
'termsId' column/code is also unnecessary and can be safely removed from
its associated 'WCGlobalAttributeModel' table.

ParaskP7 committed:
Cleanup: Remove unused product attribute related store/rest/sql code

ParaskP7 committed:
Test: Refactor wc global attribute mapper test

ParaskP7 committed:
Test: Refactor wc global attribute store test

ParaskP7 committed:
Test: Refactor wc product attribute test fixtures

ParaskP7 committed:
Refactor: Make loadCachedStoreAttributes a suspend function

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Test: Move attributes test fixtures and store test to fluxc-plugin

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Migrate: 'WCOrderStatusModel' from wellsql to room

FYI: This change includes but is not limited to the:
- Entity Introduction
- DAO Introduction (With Test)
- WCAndroidDatabase Migration
- Code Adjustments (Mapper, Store)
- Test Adjustments (Mapper, Store)

ParaskP7 committed:
Refactor: Simplify wc global attribute store fetch store attributes fun

PS: It is quite hard to make sense of all those chained scope function
calls.

ParaskP7 committed:
Db: Drop 'WCGlobalAttributeModel' table from 'wp-fluxc'

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCGlobalAttributeStore.kt:
```diff
@@ -8,48 +8,41 @@ import org.wordpress.android.fluxc.network.rest.wpcom.wc.WooError
+                }
+            }
+
+    suspend fun loadCachedStoreAttributes(
```
All good here, there is no need for backward compatibility here, this was tested and everything is working as expected. ✅

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/attribute/WCGlobalAttributeModel.kt:
```diff
@@ -1,26 +1,19 @@
-    override fun getId() = id
-}
+    val siteId: LocalId = LocalId(0),
+    val remoteId: RemoteId = RemoteId(0),
```
True, this is a good idea, thanks bot! 👍

ParaskP7 replied:
Done: d24b0c7cbf6cd71d148bde6e43e31a95cc0adc4d + 3c299858e0d28555b73d582115f3520600afc4e1 ✅

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/GlobalAttributesDao.kt:
```diff
@@ -0,0 +1,36 @@
+    @Transaction
+    open suspend fun replaceAllForSite(siteId: LocalId, attributes: List<WCGlobalAttributeModel>) {
+        deleteAttributesForSite(siteId)
+        upsertAttributes(attributes)
```
Nah, this is a far-fetch scenario that isn't going to happen. ✅

ParaskP7 committed:
Refactor: Remove unnecessary default values on 'WCGlobalAttributeModel'

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14464#discussion_r2273455708

ParaskP7 committed:
Refactor: Remove unnecessary default values on 'WCAttributeTermModel'

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14464#discussion_r2273455708

ParaskP7 commented:
I swear @malinajirka, with Wojtek AFK from tomorrow, I just used `@woocommerce/android-developers` and it randomly assigned you... 😊 😅 🫣

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/attribute/WCGlobalAttributeMapper.kt:
```diff
@@ -12,8 +13,8 @@ class WCGlobalAttributeMapper @Inject constructor() {
         WCGlobalAttributeModel(
-                remoteId = id?.toIntOrNull() ?: 0,
-                localSiteId = site.id,
+                remoteId = RemoteId(id?.toLongOrNull() ?: 0),
```
This is true, but ignoring this for now to not complicate this migration, and tbh I don't think such thing will happen, not at least to often, it is more a code safety instruction to avoid nullable values than anything else. 🤷

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCGlobalAttributeStore.kt:
```diff
@@ -8,48 +8,41 @@ import org.wordpress.android.fluxc.network.rest.wpcom.wc.WooError
 
 @Singleton
-class WCGlobalAttributeStore @Inject constructor(
+class WCGlobalAttributeStore @Inject internal constructor(
```
Nah, this is done on purpose and testing, shouldn't affect any consumers whatsoever, but thanks bot! ✅

malinajirka commented:
> I swear @malinajirka, with Wojtek AFK from tomorrow, I just used @woocommerce/android-developers and it randomly assigned you... 😊 😅 🫣

Haha 🤣 , no worries at all. I'm adding it to my todo list, but since I was AFK yesterday, there is quite a lot there. I'll review it by the end of tomorrow the latest.

ParaskP7 commented:
> Haha 🤣 , no worries at all.

🤣

> I'm adding it to my todo list, but since I was AFK yesterday, there is quite a lot there.

I bet @malinajirka , hoping all is good! 🫂

> I'll review it by the end of tomorrow the latest.

This is not urgent, not at all, please review it when you get some extra space to breath... 🙏

malinajirka reviewed (approved):
Thanks @ParaskP7! Great job, I've reviewed the code and tested the changes. 

I've encountered a small issue one time, but it's very likely unrelated to these changes and I haven't been able to reproduce it since. When I created a new variable product and added 3 attributes, the `generate variations` button would tell me there are no attributes to generate the combinations from. However, when I left that screen and entered it again, I could generate them without issues. I tried to replicate it multiple times without success.

I believe this is good to merge :shipit:. Leaving it open in case the above steps would give you some idea.

ParaskP7 commented:
👋 @malinajirka and thanks for reviewing/testing the changes! 🙇 ❤️ 🚀 

> I've encountered a small issue one time, but it's very likely unrelated to these changes and I haven't been able to reproduce it since. When I created a new variable product and added 3 attributes, the generate variations button would tell me there are no attributes to generate the combinations from. However, when I left that screen and entered it again, I could generate them without issues. I tried to replicate it multiple times without success.

Yea, I think I know what you mean, not sure if that's the same, but me not being experienced enough with this `variable product` flow overall, I too got confused many time. I went back and forward testing the flows on `trunk` and on this `branch` many times, just to make sure I am not introducing (somehow) any regressions.

On that specific flow what I noticed (and might hep) is the below:
1. When you change the product type from the usual `Physical product` to a `Variable product` and then go through the `Add variations` flow, at the very end, after clicking `NEXT` a couple of time, you get to the `GENERATE VARIATION` screen, very intuitive.
2. At this point the `Add variations` screen get transformed to another screen, the `Variations attributes` screen. 
3. Now, if you go and try the same flows as above (1.) you no longer have any `NEXT` buttons when you. add a new attribute, you can only go back on the product's main page
4. Now, if you click the `SAVE` button, only then you could `Generate all variations` via the `Variations` screen. Otherwise, if you don't save, and although you have already added the new variations attributes, when you try and `Generate all variations` via the `Variations` screen, you'll get the `No variations to generate` result.

Maybe this is what you've experienced above, you just forgot to click `SAVE`, wdyt? 🤷 

---

Btw, I like the fact that we're talking about this, I wanted to, because this might be an opportunity to further discuss this specific flow. I am pretty sure I can't be the only one who's confused by how this works, and I think that our users, unless they too tried it a few time, they'll be confused as well. 🤔

Let me know your thoughts! 🙏

> I believe this is good to merge :shipit:. Leaving it open in case the above steps would give you some idea.

:shipit:

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-global-attribute-model-to-room

# Conflicts:
#	libs/fluxc-plugin/schemas/org.wordpress.android.fluxc.persistence.WCAndroidDatabase/56.json
#	libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/di/WCDatabaseModule.kt
#	libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/WCAndroidDatabase.kt

malinajirka commented:
> Btw, I like the fact that we're talking about this, I wanted to, because this might be an opportunity to further discuss this specific flow. I am pretty sure I can't be the only one who's confused by how this works, and I think that our users, unless they too tried it a few time, they'll be confused as well. 🤔

Haha, I actually raised this discussion earlier this morning when I saw a user review - I pinged you on Slack.

> Maybe this is what you've experienced above, you just forgot to click SAVE, wdyt? 🤷

Yep, I'm almost certain this was it. Thanks for the clear steps!

ParaskP7 commented:
> Haha, I actually raised this discussion earlier this morning when I saw a user review - I pinged you on Slack.

Awesome, thanks a lot! 🙇 💯 🥇 

> Yep, I'm almost certain this was it. Thanks for the clear steps!

Thank YOU! ❤️

ParaskP7 committed:
Migrate: 'WCOrderStatusModel' from wellsql to room (v2)

Due to cbac12ffad281b00824872ee8117e8aa2fd6b141 a merge conflict
appeared which got solved via 07b967af3d447cd67939e3d286846c37c131af8e,
with the main 'WCAndroidDatabase' including:

- const val WC_DATABASE_VERSION = 57
- AutoMigration(from = 56, to = 57),

As well as accepting the current version of '56.json' and creating this
new one, the '57.json' because the database version for this exact
change has now been updated to '57' (from '56').
